File: tuan05/n_quan_hau.jl
function solve_n_queens()
    print("Nhap n: ")
    n = parse(Int, readline())

    so_nut = 0
    so_loi_giai = 0
    loi_giai_dau = Int[]

    x = zeros(Int, n)
    cot = falses(n)
    cheo1 = falses(2 * n)
    cheo2 = falses(2 * n)

    function dat_hau(i)
        so_nut += 1

        if i > n
            so_loi_giai += 1
            if so_loi_giai == 1
                loi_giai_dau = copy(x)
            end
            return
        end

        for j in 1:n
            if !cot[j] && !cheo1[i + j] && !cheo2[i - j + n]
                x[i] = j
                cot[j] = cheo1[i + j] = cheo2[i - j + n] = true

                dat_hau(i + 1)

                cot[j] = cheo1[i + j] = cheo2[i - j + n] = false
            end
        end
    end

    dat_hau(1)

    println("So nut da duyet: ", so_nut)
    println("So loi giai: ", so_loi_giai)
    println("Ban co cua loi giai dau tien:")

    if so_loi_giai > 0
        for r in 1:n
            for c in 1:n
                if loi_giai_dau[r] == c
                    print("Q ")
                else
                    print(". ")
                end
            end
            println()
        end
    else
        println("Khong co loi giai hop le.")
    end
end

solve_n_queens()

File: tuan05/n_quan_hau.cpp
#include <iostream>
#include <vector>

using namespace std;

int n;
int so_nut = 0;
int so_loi_giai = 0;
vector<int> x;
vector<int> loi_giai_dau;
vector<bool> cot, cheo1, cheo2;

void dat_hau(int i) {
    so_nut++; // Moi lan goi de quy tinh la mot nut duoc duyet[cite: 1]

    if (i == n) {
        so_loi_giai++;
        if (so_loi_giai == 1) {
            loi_giai_dau = x;
        }
        return;
    }

    for (int j = 0; j < n; j++) {
        if (!cot[j] && !cheo1[i + j] && !cheo2[i - j + n]) {
            x[i] = j;
            cot[j] = cheo1[i + j] = cheo2[i - j + n] = true;

            dat_hau(i + 1);

            // Hoan tac trang thai[cite: 1]
            cot[j] = cheo1[i + j] = cheo2[i - j + n] = false;
        }
    }
}

int main() {
    cout << "Nhap n: ";
    if (!(cin >> n) || n <= 0) return 0;

    x.assign(n, 0);
    cot.assign(n, false);
    cheo1.assign(2 * n, false);
    cheo2.assign(2 * n, false);

    dat_hau(0);

    cout << "So nut da duyet: " << so_nut << "\n";
    cout << "So loi giai: " << so_loi_giai << "\n";
    cout << "Ban co cua loi giai dau tien:\n";

    if (so_loi_giai > 0) {
        for (int r = 0; r < n; r++) {
            for (int c = 0; c < n; c++) {
                if (loi_giai_dau[r] == c) cout << "Q ";
                else cout << ". ";
            }
            cout << "\n";
        }
    } else {
        cout << "Khong co loi giai hop le.\n";
    }

    return 0;
}